File: rotate-list.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x, next=None):
        self.val = x
        self.next = next


class Solution(object):
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if head is None:
            return
        if head.next is None:
            return head
        pre = ListNode(None)
        pre.next = head
        cur = head
        l = 0
        while cur:
            l += 1
            if cur is not head:
                cur.prev = pre
            cur = cur.next
            pre = pre.next
        cur = pre
        head.prev = cur
        for _ in range(k % l):
            cur.prev.next = None
            cur.next = head
            head = cur
            cur = cur.prev
        return head


def test():
    head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
    head = Solution().rotateRight(head, 4)
    l = []
    while head:
        l.append(head.val)
        head = head.next
    assert l == [2, 3, 4, 5, 1]
    head = ListNode(1)
    head = Solution().rotateRight(head, 4)
    l = []
    while head:
        l.append(head.val)
        head = head.next
    assert l == [1]

if __name__ == '__main__':
    test()

File: algorithms/medium/rotate-list.c
#include <stdio.h>
#include <stdlib.h>

struct ListNode {
    int val;
    struct ListNode *next;
};
struct ListNode *head;

struct ListNode* newNode(int val, struct ListNode *next) {
    struct ListNode *node = malloc(sizeof(struct ListNode));
    node->val = val;
    node->next = next;
    return node;
}

struct ListNode* rotateRight(struct ListNode* head, int k){
    struct ListNode *cur = head, *tmp;
    int len = 1;
    if (head == NULL || head->next == NULL)
        return head;
    while (cur->next != NULL) {
        len++;
        cur = cur->next;
    }
    k = k % len;
    if (k == 0)
        return head;
    k = len - k -1;
    cur->next = head;
    cur = head;
    while (k-- > 0) {
        cur = cur->next;
    }
    tmp = cur->next;
    cur->next = NULL;
    return tmp;
}

int main() {
    struct ListNode *head;
    head = newNode(1, newNode(2, newNode(3, newNode(4, newNode(5, NULL)))));
    head = rotateRight(head, 2);
    while (head != NULL) {
        printf("%d ", head->val);
        head = head->next;
    }
    printf("\n");

    head = newNode(0, newNode(1, newNode(2, NULL)));
    head = rotateRight(head, 4);
    while (head != NULL) {
        printf("%d ", head->val);
        head = head->next;
    }
    printf("\n");
}
